### PR TITLE
Tsingmicro add txda backend to vllm-fl plugin.

### PR DESCRIPTION
**Pull request by tsingmicro-public-e**

### CODE REVIEW

File: vllm_fl/dispatch/backends/vendor/txda/register_ops.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        # ),
+        # # Rotary Embedding
+        # OpImpl(
+        #     op_name="rotary_embedding",
+        #     impl_id="vendor.txda",
+        #     kind=BackendImplKind.VENDOR,
+        #     fn=_bind_is_available(backend.rotary_embedding, is_avail),
+        #     vendor="txda",
+        #     priority=BackendPriority.VENDOR,
+        # ),
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/register_ops.py]: Why are we dropping the other implementations?

File: vllm_fl/dispatch/backends/vendor/txda/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2026 BAAI. All rights reserved.
+
+"""
+Txda (Tsingmicro) backend for vllm-plugin-FL dispatch.
+"""
+
+from .txda import TxdaBackend
+
+__all__ = ["TxadBackend"]
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/__init__.py]: TxdaBackend typo

File: vllm_fl/dispatch/backends/vendor/txda/register_ops.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2026 BAAI. All rights reserved.
+
+"""
+METAX backend operator registrations.
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/register_ops.py]:
```suggestion
TsingMicro backend operator registrations.
```

File: vllm_fl/dispatch/backends/vendor/txda/register_ops.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+This module registers all VENDOR (METAX) implementations.
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/register_ops.py]:
```suggestion
This module registers all VENDOR (TsingMicro) implementations.
```

File: vllm_fl/dispatch/backends/vendor/txda/register_ops.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    def wrapper(*args, **kwargs):
+        return fn(*args, **kwargs)
+
+    wrapper._is_available = is_available_fn
+    return wrapper
+
+
+def register_builtins(registry) -> None:
+    """
+    Register all METAX (VENDOR) operator implementations.
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/register_ops.py]:
```suggestion
    Register all TsingMicro (VENDOR) operator implementations.
```

File: vllm_fl/dispatch/backends/vendor/txda/txda.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+This backend provides operator implementations for Tsingmiocro Txda NPUs.
+"""
+
+from __future__ import annotations
+
+from typing import Optional, Union
+
+import torch
+
+# from vllm_fl.dispatch.backends.flaggems import FlagGemsBackend
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/txda.py]: Remove this line.

File: vllm_fl/dispatch/backends/vendor/txda/txda.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    @property
+    def vendor(self) -> Optional[str]:
+        return "txda"
+
+    def is_available(self) -> bool:
+        """Check if Txda hardware and libraries are available."""
+        if TxdaBackend._available is None:
+            try:
+                # Check for torch_npu (Txda PyTorch extension)
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/txda.py]:
```suggestion
                # Check for torch_txda (Txda PyTorch extension)
```

File: vllm_fl/dispatch/backends/vendor/txda/txda.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                import torch_txda
+
+                # Check if NPU device is available
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/txda.py]:
```suggestion
                # Check if TsingMicro txda device is available
```

File: vllm_fl/dispatch/backends/vendor/txda/txda.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                if torch.txda.is_available() and torch.txda.device_count() > 0:
+                    TxdaBackend._available = True
+                else:
+                    TxdaBackend._available = False
```
Comment (tengqm) [inline vllm_fl/dispatch/backends/vendor/txda/txda.py]: It is weird that we change class properties in an instance method.
You may want to annotate this method with `@classmethod`.

File: vllm_fl/distributed/device_communicators/flagcx.py [hunk 10]
```diff
@@ -72,7 +72,8 @@ def __init__(
             ### TODO(lms): simplify it
             if library_path is None:
                 flagcx_path = os.getenv('FLAGCX_PATH')
-                library_path=os.path.join(flagcx_path, "build/lib/libflagcx.so")
+                #library_path=os.path.join(flagcx_path, "libflagcx.so") # rcy fix
+                library_path= "/usr/local/kuiper/lib/libflagcx.so"
```
Comment (tengqm) [inline vllm_fl/distributed/device_communicators/flagcx.py]: We are using hard-coded path rather than using the environment variable?

File: vllm_fl/ops/fla/fused_recurrent.py [hunk 11]
```diff
@@ -6,13 +6,13 @@
 
 from vllm_fl.utils import use_flaggems_op
 
-if use_flaggems_op("fused_recurrent_gated_delta_rule_fwd"):
+if True:
```
Comment (tengqm) [inline vllm_fl/ops/fla/fused_recurrent.py]: Why this change?

File: vllm_fl/platform.py [hunk 12]
```diff
@@ -46,21 +46,18 @@ class PlatformFL(Platform):
     dispatch_key = device_info.dispatch_key
     torch_device_fn = device_info.torch_device_fn
     vendor_name = device_info.vendor_name
-    ray_device_key: str = "GPU"
+    ray_device_key: str = "flagos"
     dist_backend: str = "flagcx" if "FLAGCX_PATH" in os.environ else "nccl"
+    device_control_env_var: str = "TXDA_VISIBLE_DEVICES"
```
Comment (tengqm) [inline vllm_fl/platform.py]: This is not a platform specific class ...
This property should not live here?

**Review by tengqm — COMMENTED** [review]
Please label this PR with `[WIP]` if it is still a work in progress,
in case it gets merged accidentally.

File: vllm_fl/platform.py [hunk 13]
```diff
@@ -46,21 +46,18 @@ class PlatformFL(Platform):
     dispatch_key = device_info.dispatch_key
     torch_device_fn = device_info.torch_device_fn
     vendor_name = device_info.vendor_name
-    ray_device_key: str = "GPU"
+    ray_device_key: str = "flagos"
```
Comment (cyber-pioneer) [inline vllm_fl/platform.py]: why modify here?

File: vllm_fl/platform.py [hunk 14]
```diff
@@ -46,21 +46,18 @@ class PlatformFL(Platform):
...
     dist_backend: str = "flagcx" if "FLAGCX_PATH" in os.environ else "nccl"
+    device_control_env_var: str = "TXDA_VISIBLE_DEVICES"
     ### TODO(lms): dispatch device_control_env_var
     # device_control_env_var: str = "CUDA_VISIBLE_DEVICES"
 
     def is_cuda_alike(self) -> bool:
         """Stateless version of [torch.cuda.is_available][]."""
-        if self.vendor_name == "iluvatar":
```
Comment (cyber-pioneer) [inline vllm_fl/platform.py:70]: why remove here?
